java/sage/MiniClientSageRenderer.java

```diff
@@ -1328,6 +1328,33 @@ private boolean checkForAudioOutputChangeUpdates()
     return true;
   }
 
+  private boolean checkForMenuHintUpdates()
+  {
+    String hintText;
+    synchronized (menuHint) {
+      hintText=menuHint.format();
+      menuHint.clear();
+    }
+    if (hintText!=null)
```

Comment:
That makes a lot of sense.  I tend to be very lenient or very strict with
locking :)  In this case because it was "hints" I didn't add any locking,
since my feeling was that the chances of the player getting the wrong hint
would be minimal, and likely short lived (ie, next hint with a menu or
popup would rectify it).  But, your logic is pretty good and minimizes the
locking until it is needed.  If I get a chance I'll look at refactoring the
code based on your explanation.

Thanks,
Sean.

On Tue, 19 Jan 2016 at 19:28 Jeffrey Kardatzke <notifications@github.com>
wrote:

> In java/sage/MiniClientSageRenderer.java
> <https://github.com/google/sagetv/pull/71#discussion_r50198533>:
>
> > @@ -1328,6 +1328,33 @@ private boolean checkForAudioOutputChangeUpdates()
> >      return true;
> >    }
> >
> > +  private boolean checkForMenuHintUpdates()
> > +  {
> > +    String hintText;
> > +    synchronized (menuHint) {
> > +      hintText=menuHint.format();
> > +      menuHint.clear();
> > +    }
> > +    if (hintText!=null)
>
> Calling hasHint() outside of the synchronized block should be perfectly
> fine. The state may get updated simultaneously with that check..but it
> won't get cleared, so the return value from hasHint() is still valid
> outside the sync block. I did want to avoid grabbing the lock every time
> this checks. The purpose of the synchronization was two-fold: 1. Ensure
> that all fields of MenuHint are from a single call to setMenuHint when
> creating the format(). This is achieved by synchronizing the call to
> format() as well as setMenuHint(). 2. Ensure that when you clear() the
> MenuHint structure, that you are clearing what you already took the format
> for (so you don't get the format() from a prior setMenuHint(), then a new
> hint is set, and then you clear it and never end up sending the new hint).
> This is achieved by having format() and clear() in the same sync block.
> It's fine the way you have it now...it's not like one extra lock for each
> rendering cycle is going to hurt anything...there's tons of locking already
> that happens on each render cycle. :) I'll approve it the way it is, I just
> explained things above so you understood my line of thinking. Of course my
> solution suffers from only focusing on the singular usage that currently is
> there...yours is more robust for other usages that may come in later. :)
> … <#msg-f:1523842769132898788_>
> On Tue, Jan 19, 2016 at 3:56 PM, Sean Stuckless <notifications@github.com>
> wrote: In java/sage/MiniClientSageRenderer.java <#71 (comment)
> <https://github.com/google/sagetv/pull/71#discussion_r50195941>>: > @@
> -1328,6 +1328,33 @@ private boolean checkForAudioOutputChangeUpdates() >
> return true; > } > > + private boolean checkForMenuHintUpdates() > + { > +
> String hintText; > + synchronized (menuHint) { > +
> hintText=menuHint.format(); > + menuHint.clear(); > + } > + if
> (hintText!=null) That's a good point, but, calling hasHint outside of the
> syncronized block seems to defeat the purpose of actually synchronizing in
> the first place. I think that format() should return null if there is no
> hint, and that way the entire acess to menuHint remains within a sync
> block. Otherwise, I have to put al the code if if(menuHint.hasHint())
> inside of a sync block and my feeling was that would be holding the lock
> for too long. — Reply to this email directly or view it on GitHub <
> https://github.com/google/sagetv/pull/71/files#r50195941>.
> -- Jeffrey Kardatzke jkardatzke@google.com Google, Inc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/sagetv/pull/71/files#r50198533>.
>
-- 

Sent from Android